applications/tari_base_node/src/parser.rs

```diff
@@ -747,3 +740,76 @@ impl Parser {
         self.command_handler.lock().await.list_reorgs();
     }
 }
+
+use std::str::SplitWhitespace;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error("{name} {reason}")]
+struct ArgsError {
+    name: &'static str,
+    reason: ArgsReason,
+}
+
+impl ArgsError {
+    pub fn new(name: &'static str, reason: ArgsReason) -> Self {
+        Self { name, reason }
+    }
+}
+
+#[derive(Debug, Error)]
+enum ArgsReason {
+    #[error("argument required")]
+    Required,
+    #[error("argument can't be parsed: {details}")]
+    NotParsed { details: String },
+}
+
+struct Args<'a> {
+    splitted: Peekable<SplitWhitespace<'a>>,
+}
+
+impl<'a> Args<'a> {
+    fn split(s: &'a str) -> Self {
+        Self {
+            splitted: s.split_whitespace().peekable(),
+        }
+    }
+
+    fn shift_one(&mut self) {
+        self.splitted.next();
+    }
+
+    // TODO: It have to return error if a value provided,
+    // but can''t be parsed
+    fn take_node_id(&mut self) -> Option<NodeId> {
+        self.splitted
+            .next()
+            .and_then(parse_emoji_id_or_public_key_or_node_id)
+            .map(either_to_node_id)
+    }
+
+    fn try_take_next<T>(&mut self, name: &'static str) -> Result<Option<T>, ArgsError>
+    where
+        T: FromStr,
+        T::Err: ToString,
+    {
+        match self.splitted.peek().map(|s| s.parse()) {
```

Comment:
Why `peek` and not `next`? Peek wont progress the iterator so try_take_next will always return the same value